### PR TITLE
Authorization 2

### PR DESCRIPTION
**Pull request by CorneliaKelinske**
@alanvardy final review please

I have implemented authorization for a number of user and upload functions.
The corresponding tests have been updated and tests to account for a scenario where the user is logged in but not an admin have been added.

Thank you!

### CODE REVIEW

File: lib/the_little_thinkers_space/content/upload.ex [hunk 1]
```diff
@@ -39,8 +39,9 @@ defmodule TheLittleThinkersSpace.Content.Upload do
     |> validate_inclusion(:file_type, @valid_file_types, message: "Wrong file type!")
   end
 
-  def authorize(action, %{role: "Admin"} = _user, _upload) when action in [:new, :create], do: :ok
-  def authorize(action, %{role: "TheLittleThinker"} = _user, _upload) when action in [:new, :create], do: :ok
-  def authorize(_action, _user, _post), do: :error
+  def authorize(_, %{role: "Admin"} = _user, _upload), do: :ok
+  def authorize(_, %{role: "TheLittleThinker"} = _user, _upload), do: :ok
```
Comment (alanvardy) [inline lib/the_little_thinkers_space/content/upload.ex:43]:
```suggestion
  def authorize(actions, %User{id: id, role: "TheLittleThinker"}, %Upload{user_id: id) when actions in [:edit, :update, :delete], do: :ok
```

Comment (alanvardy) [reply lib/the_little_thinkers_space/content/upload.ex:43]: This verifies that there is a user and an upload, and that the upload belongs to the user.

File: lib/the_little_thinkers_space/content/upload.ex [hunk 2]
```diff
@@ -39,8 +39,9 @@ defmodule TheLittleThinkersSpace.Content.Upload do
...
 
+  def authorize(action, _user, _upload) when action in [:show, :index], do: :ok
+  def authorize(_action, _user, _upload), do: :error
```
Comment (alanvardy) [inline lib/the_little_thinkers_space/content/upload.ex:46]: :+1:

File: lib/the_little_thinkers_space/content/upload.ex [hunk 3]
```diff
@@ -39,8 +39,9 @@ defmodule TheLittleThinkersSpace.Content.Upload do
     |> validate_inclusion(:file_type, @valid_file_types, message: "Wrong file type!")
   end
 
-  def authorize(action, %{role: "Admin"} = _user, _upload) when action in [:new, :create], do: :ok
-  def authorize(action, %{role: "TheLittleThinker"} = _user, _upload) when action in [:new, :create], do: :ok
-  def authorize(_action, _user, _post), do: :error
+  def authorize(_, %{role: "Admin"} = _user, _upload), do: :ok
+  def authorize(_, %{role: "TheLittleThinker"} = _user, _upload), do: :ok
 
+  def authorize(action, _user, _upload) when action in [:show, :index], do: :ok
```
Comment (alanvardy) [inline lib/the_little_thinkers_space/content/upload.ex:45]: A non-logged in user can get past this as the user argument could just be nil here?

File: lib/the_little_thinkers_space_web/controllers/user_controller.ex [hunk 4]
```diff
@@ -4,28 +4,46 @@ defmodule TheLittleThinkersSpaceWeb.UserController do
...
+    admin = conn.assigns.current_user
+
+    with :ok <- Bodyguard.permit(User, :create, admin) do
+      changeset = Accounts.change_registration(%User{}, %{})
+      render(conn, "new.html", changeset: changeset)
+    else
+      {:error, :unauthorized} ->
+        conn
+        |> put_flash(:error, "You are not allowed to do this!")
+        |> redirect(to: Routes.upload_path(conn, :index))
```
Comment (alanvardy) [inline lib/the_little_thinkers_space_web/controllers/user_controller.ex:24]: You can replace all this else code in all the controllers with the ActionFallback macro
https://hexdocs.pm/bodyguard/readme.html#controllers

Comment (alanvardy) [reply lib/the_little_thinkers_space_web/controllers/user_controller.ex:24]: Just kick them all back to the login page I figure

File: test/the_little_thinkers_space_web/controllers/upload_controller_test.exs [hunk 5]
```diff
@@ -116,39 +133,60 @@ defmodule TheLittleThinkersSpaceWeb.UploadControllerTest do
...
+      user = UserHelper.user_fixture()
+      conn = UserHelper.login_user(conn, user)
+      conn = post(conn, Routes.upload_path(conn, :create), upload: @create_video_attrs)
+      assert html_response(conn, 302) =~ "You are being <a href=\"/uploads\">redirected</a>."
+    end
   end
 
   describe "edit upload" do
-    setup [:create_upload]
+    setup [:create_admin_upload]
```
Comment (alanvardy) [inline test/the_little_thinkers_space_web/controllers/upload_controller_test.exs:152]: :+1:

File: lib/the_little_thinkers_space/accounts/user.ex [hunk 6]
```diff
@@ -48,4 +49,8 @@ defmodule TheLittleThinkersSpace.Accounts.User do
         changeset
     end
   end
+
+  def authorize(_, %{role: "Admin"} = _user, _upload), do: :ok
+  def authorize(action, _role, _user) when action in [:index], do: :ok
```
Comment (alanvardy) [inline lib/the_little_thinkers_space/accounts/user.ex:54]:
```suggestion
  def authorize(:index, _role, _user), do: :ok
```

File: lib/the_little_thinkers_space_web/controllers/user_controller.ex [hunk 7]
```diff
@@ -4,28 +4,46 @@ defmodule TheLittleThinkersSpaceWeb.UserController do
...
     users = Accounts.list_users()
     render(conn, "index.html", users: users)
   end
 
   def new(conn, _params) do
-    changeset = Accounts.change_registration(%User{}, %{})
-    render(conn, "new.html", changeset: changeset)
+    admin = conn.assigns.current_user
+
+    with :ok <- Bodyguard.permit(User, :create, admin) do
```
Comment (alanvardy) [inline lib/the_little_thinkers_space_web/controllers/user_controller.ex:17]: `admin` should be going in as the first argument right?

Comment (CorneliaKelinske) [reply lib/the_little_thinkers_space_web/controllers/user_controller.ex:17]: Thank you! I had a look at this, but it is correct :)

**Pull request merged**